src/plugins/data_source_management/server/saved_objects/data_source_premission_client_wrapper.ts

```diff
@@ -0,0 +1,231 @@
+ * DataSourceAdmin user has all permissions.
+ * If manageableBy is all, any user has permissions.
+ * If manageableBy is none, any user has no permissions.
+ * If manageableBy is dashboard_admin, only OSD admin has permissions.
+ */
+export class DataSourcePermissionClientWrapper {
+  constructor(private manageableBy: string) {}
+
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const { isDashboardAdmin, isDataSourceAdmin } = getWorkspaceState(wrapperOptions.request) || {};
+    if (isDataSourceAdmin) {
+      return wrapperOptions.client;
+    }
+    // If isDashboardAdmin is undefined / true, the user will be dashboard admin
+    const isDashboardAdminRequest = isDashboardAdmin !== false;
+
+    const canSkipWrapper =
+      this.manageableBy === ManageableBy.All ||
+      (this.manageableBy === ManageableBy.DashboardAdmin && isDashboardAdminRequest);
+
+    const createWithManageableBy = async <T = unknown>(
+      type: string,
+      attributes: T,
+      options?: SavedObjectsCreateOptions
+    ) => {
+      if (DATA_SOURCE_SAVED_OBJECT_TYPE === type && !canSkipWrapper) {
+        throw this.generatePermissionError();
+      }
+      return await wrapperOptions.client.create(type, attributes, options);
+    };
+
+    const bulkCreateWithManageableBy = async <T = unknown>(
+      objects: Array<SavedObjectsBulkCreateObject<T>>,
+      options?: SavedObjectsCreateOptions
+    ): Promise<SavedObjectsBulkResponse<T>> => {
+      if (canSkipWrapper) {
+        return await wrapperOptions.client.bulkCreate(objects, options);
+      }
+      const disallowedSavedObjects: Array<SavedObjectsBulkCreateObject<T>> = [];
+      const allowedSavedObjects: Array<SavedObjectsBulkCreateObject<T>> = [];
+
+      objects.forEach((item) => {
+        if (DATA_SOURCE_SAVED_OBJECT_TYPE === item.type) {
+          disallowedSavedObjects.push(item);
+          return;
+        }
+
+        allowedSavedObjects.push(item);
+        return;
+      });
+
+      const bulkCreateResult = await wrapperOptions.client.bulkCreate(allowedSavedObjects, options);
+
+      // Merge the data source saved objects and real client bulkCreate result.
+      return {
+        saved_objects: [
+          ...(bulkCreateResult?.saved_objects || []),
+          ...disallowedSavedObjects.map((item) => ({
+            ...item,
+            error: {
+              ...this.generatePermissionError().output.payload,
+              metadata: { isNotOverwritable: true },
+            },
+          })),
+        ],
+      } as SavedObjectsBulkResponse<T>;
+    };
+
+    const updateWithManageableBy = async <T = unknown>(
+      type: string,
+      id: string,
+      attributes: Partial<T>,
+      options: SavedObjectsUpdateOptions = {}
+    ): Promise<SavedObjectsUpdateResponse<T>> => {
+      if (
+        DATA_SOURCE_SAVED_OBJECT_TYPE !== type ||
+        (this.manageableBy === ManageableBy.DashboardAdmin && isDashboardAdminRequest)
```

Comment:
```suggestion
        canSkipWrapper
```

Reply:
When manageableBy is all, we should also check whether the options have the workspaces field. Only dashboard admin can update the workspaces field. So we should not skip the wrapper directly.